marlowe/src/Language/Marlowe/Scripts.hs

```diff
@@ -276,24 +291,73 @@ mkMarloweValidator
         find (\TxInInfo{txInInfoOutRef} -> txInInfoOutRef == txOutRef) txInfoInputs
     findOwnInput _ = Nothing
 
-    -- The inputs being spent by this script.
     -- [Marlowe-Cardano Specification: "2. Single Marlowe script input".]
+    -- The inputs being spent by this script, and whether other validators are present.
     ownInput :: TxInInfo
-    ownInput@TxInInfo{txInInfoResolved=TxOut{txOutAddress=ownAddress}} =
+    noOthers :: Bool
+    (ownInput@TxInInfo{txInInfoResolved=TxOut{txOutAddress=ownAddress}}, noOthers) =
         case findOwnInput ctx of
-            Just ownTxInInfo ->
-                case filter (sameValidatorHash ownTxInInfo) (txInfoInputs scriptContextTxInfo) of
-                    [i] -> i
-                    _   -> traceError "w" -- Multiple Marlowe contract inputs with the same address are forbidden.
+            Just ownTxInInfo -> examineScripts (sameValidatorHash ownTxInInfo) Nothing True (txInfoInputs scriptContextTxInfo)
             _ -> traceError "x" -- Input to be validated was not found.
 
+    -- Check for the presence of multiple Marlowe validators or other Plutus validators.
+    examineScripts
+      :: (ValidatorHash -> Bool)  -- Test for this validator.
+      -> Maybe TxInInfo           -- The input for this validator, if found so far.
+      -> Bool                     -- Whether no other validator has been found so far.
+      -> [TxInInfo]               -- The inputs remaining to be examined.
+      -> (TxInInfo, Bool)         -- The input for this validator and whehter no other validators are present.
+    -- This validator has not been found.
+    examineScripts _ Nothing _ [] = traceError "x"
+    -- This validator has been found, and other validators may have been found.
+    examineScripts _ (Just self) noOthers [] = (self, noOthers)
+    -- This validator has not yet been found.
+    examineScripts f Nothing noOthers (tx@TxInInfo{txInInfoResolved=TxOut{txOutAddress=Ledger.Address (ScriptCredential vh) _}} : txs)
+      -- This validator is found, but continue looking for other validators if necessary.
+      | f vh = if noOthers then examineScripts f (Just tx) noOthers txs else (tx, False)
+      -- Another validator is found, but continue looking for this validator.
+      | otherwise = examineScripts f Nothing False txs
+    -- This validator has been found, but other validators have not yet been found.
+    examineScripts f (Just self) _ (TxInInfo{txInInfoResolved=TxOut{txOutAddress=Ledger.Address (ScriptCredential vh) _}} : _)
+      -- This validator is found a second time.
+      | f vh = traceError "w"
+      -- Another validator is found, so there is no need to look further.
+      | otherwise = (self, False)
```

Comment:
```suggestion
    -- Found both this validator and another script, so we short-cut.
    examineScripts _ (Just self) False _ = (self, False)
     -- Found one script.
    examineScripts f mSelf noOthers (tx@TxInInfo{txInInfoResolved=TxOut{txOutAddress=Ledger.Address (ScriptCredential vh) _}} : txs)
      -- The script is this validator.
      | f vh = case mSelf of
                 -- We hadn't found it before, so we save it in `mSelf`.
                 Nothing -> examineScripts f (Just tx) noOthers txs
                 -- We already had found this validator before
                 Just _ -> traceError "w"
      -- The script is something else, so we set `noOther` to `False`.
      | otherwise = examineScripts f mSelf False txs
```
I think the logic is sound. I found it a bit confusing that the variable is negated (`noOthers` instead of `others`), but I didn't touch that. I've also given a go at simplifying the logic by moving the shortcut (which appeared twice) to the top level and moving the matching to a case. Maybe it gives a smaller validator. On the other hand, execution time may be a bit higher. Just an idea

Reply:
@palas, you're suggested change saves 24 bytes! Change included in 57d1e34e5a8d499dd2ba4a3758fd851e7336e286.